[Flow] Function predicate declaration

panagosg7 opened the pull request:
The accepted syntax for function declarations is extended to allow
the specification of a function predicate (inferred or explicit). This
should be placed right after the function's return type (if existing,
otherwise the colon that would designate the start of the return 
type). So the return type syntax becomes:

```
FunctionReturnType :=
  Type
  Predicate
  Type Predicate

Predicate :=
  %checks
  %checks ( ConditionalExpression )
```

These changes are with regards to [this](https://github.com/facebook/flow/commit/6cbb327b7678a9c7050e7cacaa1ca2cf751c08ae) commit in the Flow repo.

danez commented on src/plugins/flow.js:
```diff
@@ -21,6 +21,45 @@ pp.flowParseTypeInitialiser = function (tok, allowLeadingPipeOrAnd) {
+  this.expectContextual("checks");
+  // Force '%' and 'checks' to be adjacent
+  if (moduloLoc.line !== checksLoc.line || moduloLoc.column !== checksLoc.column - 1) {
+    this.raise(moduloPos, "Unexpected token");
```
This line is not covered. Do you think we can have two tests that check if there are parse errors in these cases?

danez replied:
You could also use `this.unexpected(moduloPos);` here. It's a little bit shorter but otherwise isn't really important if not.

danez commented:
Thank you for your first PR to babylon. 

In order this works throughout the complete chain of babel, all the new node types (I think it is only DeclaredPredicate  and InferredPredicate) need to be added to [babel-types](https://github.com/babel/babel/blob/master/packages/babel-types/src/definitions/flow.js) and [babel-generator](https://github.com/babel/babel/blob/master/packages/babel-generator/src/generators/flow.js)
Could you do that and create a PR for babel?

I would wait with the merge till we have a PR in babel and the new flow version is released.

Great work!

panagosg7 commented:
Thanks @danez! I'll address your comments

panagosg7 commented:
@danez: This can still wait until Flow's upcoming release.

danez commented on src/plugins/flow.js:
```diff
@@ -21,6 +21,46 @@ pp.flowParseTypeInitialiser = function (tok, allowLeadingPipeOrAnd) {
+  this.expectContextual("checks");
+  // Force '%' and 'checks' to be adjacent
+  if (moduloLoc.line !== checksLoc.line || moduloLoc.column !== checksLoc.column - 1) {
+    this.unexpected(moduloPos);
```
Can you add a nice error message here? Maybe something like `"Spaces between ´%´ and ´checks´ are not allowed here."`

danez commented on src/plugins/flow.js:
```diff
@@ -21,6 +21,46 @@ pp.flowParseTypeInitialiser = function (tok, allowLeadingPipeOrAnd) {
+    this.unexpected(moduloPos);
+  }
+  if (this.match(tt.parenL)) {
+    this.next();
```
Instead of 
```js
if (this.match(tt.parenL)) {
    this.next();
```
you can do 
```js
if (this.eat(tt.parenL)) {
```

Which is doing the exact same thing.

panagosg7 commented:
Thanks for the pointers, @danez!

I will get to these, but I'm still waiting for support for function predicates to be fully merged in Flow master.

loganfsmyth commented:
I know flow supports this now, is this waiting on more changes, or has it just been forgotten?

panagosg7 commented on src/plugins/flow.js:
```diff
@@ -21,6 +21,46 @@ pp.flowParseTypeInitialiser = function (tok, allowLeadingPipeOrAnd) {
+
+pp.flowParseTypeAndPredicateInitialiser = function () {
+  let oldInType = this.state.inType;
+  this.state.inType = true;
```
It looks like `this.state.inType` needs to be `true` for `this.expect(tt.colon)` to succeed (I tried to move it further down, but other tests failed). So, `oldInType` needs to be recorded in the beginning in order to be restored before `this.flowParsePredicate()` is called (note that `flowParsePredicate` possibly parses expressions).

Bottom line is I wasn't able to perform the optimization you suggested.

panagosg7 commented:
Sorry for taking so long to get back to this.

AFAIK, there is still some function predicate code that hasn't been merged to Flow's master. Most of these changed involve the type checker, but there is also some parser work under way. For example I believe the notation of [predicate refinement](https://github.com/facebook/flow/blob/master/tests/predicates-abstract/filter.js#L5) will be updated to something more elegant. Also, there has not been an official announcement yet.

However, you are right that some support for function predicates (including `%checks`) is already in mainline and recognized by the native Flow parser, so I guess there should be support here as well.

I recently rebased my original PR, addressed @danez 's comments (including the ones in the [respective babel PR](https://github.com/babel/babel/pull/3700)) and added one more example.

leebyron commented:
Any progress on this? I'd love to see this land to unblock early use
